File: python/PetShop.py
from tabulate import tabulate

class PetShop:
    def __init__(self):
        # Inisialisasi atribut privat untuk menyimpan data produk
        self._product_ids = []          # List untuk menyimpan ID produk
        self._product_names = []       # List untuk menyimpan nama produk
        self._product_categories = []  # List untuk menyimpan kategori produk
        self._product_prices = []      # List untuk menyimpan harga produk
        self._product_photos = []      # List untuk menyimpan foto produk
        self._next_id = 1              # Variabel untuk menghasilkan ID otomatis

    # Getter untuk mendapatkan list ID produk
    def get_product_ids(self):
        return self._product_ids

    # Setter untuk mengatur ID produk pada indeks tertentu
    def set_product_id(self, index, id):
        if 0 <= index < len(self._product_ids):  # Validasi indeks
            self._product_ids[index] = id
        else:
            print("Indeks tidak valid!")  # Pesan error jika indeks tidak valid

    # Getter untuk mendapatkan list nama produk
    def get_product_names(self):
        return self._product_names

    # Setter untuk mengatur nama produk pada indeks tertentu
    def set_product_name(self, index, name):
        if 0 <= index < len(self._product_names):  # Validasi indeks
            self._product_names[index] = name
        else:
            print("Indeks tidak valid!")  # Pesan error jika indeks tidak valid

    # Getter untuk mendapatkan list kategori produk
    def get_product_categories(self):
        return self._product_categories

    # Setter untuk mengatur kategori produk pada indeks tertentu
    def set_product_category(self, index, category):
        if 0 <= index < len(self._product_categories):  # Validasi indeks
            self._product_categories[index] = category
        else:
            print("Indeks tidak valid!")  # Pesan error jika indeks tidak valid

    # Getter untuk mendapatkan list harga produk
    def get_product_prices(self):
        return self._product_prices

    # Setter untuk mengatur harga produk pada indeks tertentu
    def set_product_price(self, index, price):
        if 0 <= index < len(self._product_prices) and price >= 0:  # Validasi indeks dan harga
            self._product_prices[index] = price
        else:
            print("Harga tidak boleh negatif atau indeks tidak valid!")  # Pesan error jika tidak valid

    # Getter untuk mendapatkan list foto produk
    def get_product_photos(self):
        return self._product_photos

    # Setter untuk mengatur foto produk pada indeks tertentu
    def set_product_photo(self, index, photo):
        if 0 <= index < len(self._product_photos):  # Validasi indeks
            self._product_photos[index] = photo
        else:
            print("Indeks tidak valid!")  # Pesan error jika indeks tidak valid

    # Getter untuk mendapatkan ID berikutnya
    def get_next_id(self):
        return self._next_id

    # Method untuk menampilkan semua produk dalam bentuk tabel
    def display_products(self):
        if not self._product_ids:  # Cek apakah list produk kosong
            print("Tidak ada produk yang tersedia.")
            return
        
        # Membuat list of lists untuk data produk
        table_data = []
        for i in range(len(self._product_ids)):
            table_data.append([
                self._product_ids[i],
                self._product_names[i],
                self._product_categories[i],
                f"Rp{self._product_prices[i]}",
                self._product_photos[i]
            ])
        
        # Menampilkan tabel menggunakan tabulate
        headers = ["ID", "Nama", "Kategori", "Harga", "Foto"]
        print(tabulate(table_data, headers=headers, tablefmt="pretty"))

    # Method untuk menambahkan produk baru
    def add_product(self, name, category, price, photo):
        if price < 0:  # Validasi harga tidak boleh negatif
            print("Harga tidak boleh negatif!")
            return
        # Menambahkan data produk ke dalam list
        self._product_ids.append(self._next_id)
        self._next_id += 1  # Increment ID untuk produk berikutnya
        self._product_names.append(name)
        self._product_categories.append(category)
        self._product_prices.append(price)
        self._product_photos.append(photo)
        print("Produk berhasil ditambahkan.")

    # Method untuk memperbarui produk
    def update_product(self, id, name, category, price, photo):
        if id in self._product_ids:  # Cek apakah ID produk ada
            index = self._product_ids.index(id)  # Cari indeks produk berdasarkan ID
            # Update data produk menggunakan setter
            self.set_product_name(index, name)
            self.set_product_category(index, category)
            self.set_product_price(index, price)
            self.set_product_photo(index, photo)
            print("Produk berhasil diperbarui.")
        else:
            print(f"Produk dengan ID {id} tidak ditemukan.")  # Pesan error jika ID tidak ditemukan

    # Method untuk menghapus produk
    def delete_product(self, id):
        if id in self._product_ids:  # Cek apakah ID produk ada
            index = self._product_ids.index(id)  # Cari indeks produk berdasarkan ID
            # Hapus data produk dari semua list
            self._product_ids.pop(index)
            self._product_names.pop(index)
            self._product_categories.pop(index)
            self._product_prices.pop(index)
            self._product_photos.pop(index)
            print("Produk berhasil dihapus.")
        else:
            print(f"Produk dengan ID {id} tidak ditemukan.")  # Pesan error jika ID tidak ditemukan

    # Method untuk mencari produk berdasarkan nama
    def search_product_by_name(self, name):
        found = False  # Flag untuk menandai apakah produk ditemukan
        table_data = []  # List untuk menyimpan data produk yang ditemukan
        # Loop untuk mencari produk berdasarkan nama
        for i in range(len(self._product_names)):
            if self._product_names[i] == name:
                # Tambahkan data produk ke dalam list
                table_data.append([
                    self._product_ids[i],
                    self._product_names[i],
                    self._product_categories[i],
                    f"Rp{self._product_prices[i]}",
                    self._product_photos[i]
                ])
                found = True  # Set flag menjadi True
        
        if found:  # Jika produk ditemukan
            headers = ["ID", "Nama", "Kategori", "Harga", "Foto"]
            print(tabulate(table_data, headers=headers, tablefmt="pretty"))
        else:  # Jika produk tidak ditemukan
            print(f"Produk dengan nama '{name}' tidak ditemukan.")


def main():
    my_pet_shop = PetShop()  # Membuat objek PetShop
    while True:
        # Menampilkan menu utama
        print("\n=== SISTEM MANAJEMEN PETSHOP ===")
        print("1. Tampilkan Data Produk")
        print("2. Tambah Produk Baru")
        print("3. Perbarui Produk")
        print("4. Hapus Produk")
        print("5. Cari Produk Berdasarkan Nama")
        print("6. Keluar")
        choice = input("Pilih menu (1-6): ")  # Input pilihan menu dari pengguna

        if choice == "1":
            my_pet_shop.display_products()  # Tampilkan semua produk
        elif choice == "2":
            # Input data produk baru
            name = input("Masukkan Nama: ")
            category = input("Masukkan Kategori: ")
            price = float(input("Masukkan Harga: "))
            photo = input("Masukkan Foto: ")
            my_pet_shop.add_product(name, category, price, photo)  # Tambahkan produk baru
        elif choice == "3":
            # Input data untuk memperbarui produk
            id = int(input("Masukkan ID Produk: "))
            name = input("Masukkan Nama: ")
            category = input("Masukkan Kategori: ")
            price = float(input("Masukkan Harga: "))
            photo = input("Masukkan Foto: ")
            my_pet_shop.update_product(id, name, category, price, photo)  # Perbarui produk
        elif choice == "4":
            # Input ID produk yang akan dihapus
            id = int(input("Masukkan ID Produk: "))
            my_pet_shop.delete_product(id)  # Hapus produk
        elif choice == "5":
            # Input nama produk yang akan dicari
            name = input("Masukkan Nama Produk: ")
            my_pet_shop.search_product_by_name(name)  # Cari produk berdasarkan nama
        elif choice == "6":
            print("Terima kasih!")  # Keluar dari program
            break
        else:
            print("Pilihan tidak valid!")  # Pesan error jika pilihan tidak valid


if __name__ == "__main__":
    main()  # Jalankan program utama

File: java/PetShop.java
import java.util.ArrayList; // Mengimpor kelas ArrayList untuk menyimpan data dalam bentuk list
import java.util.List; // Mengimpor interface List untuk mendefinisikan list
import java.util.Scanner; // Mengimpor kelas Scanner untuk menerima input dari pengguna

public class PetShop {
    // Deklarasi list untuk menyimpan data produk
    private List<Integer> productIds; // List untuk menyimpan ID produk
    private List<String> productNames; // List untuk menyimpan nama produk
    private List<String> productCategories; // List untuk menyimpan kategori produk
    private List<Double> productPrices; // List untuk menyimpan harga produk
    private List<String> productPhotos; // List untuk menyimpan foto produk
    private int nextId; // Variabel untuk menghasilkan ID otomatis

    // Konstruktor untuk inisialisasi list dan variabel
    public PetShop() {
        productIds = new ArrayList<>(); // Inisialisasi list ID produk
        productNames = new ArrayList<>(); // Inisialisasi list nama produk
        productCategories = new ArrayList<>(); // Inisialisasi list kategori produk
        productPrices = new ArrayList<>(); // Inisialisasi list harga produk
        productPhotos = new ArrayList<>(); // Inisialisasi list foto produk
        nextId = 1; // Inisialisasi ID awal
    }

    public List<Integer> getProductIds() { return productIds; }
    public List<String> getProductNames() { return productNames; }
    public List<String> getProductCategories() { return productCategories; }
    public List<Double> getProductPrices() { return productPrices; }
    public List<String> getProductPhotos() { return productPhotos; }

    public void setProductName(int index, String name) { productNames.set(index, name); }
    public void setProductCategory(int index, String category) { productCategories.set(index, category); }
    public void setProductPrice(int index, double price) { productPrices.set(index, price); }
    public void setProductPhoto(int index, String photo) { productPhotos.set(index, photo); }


    // Mengambil panjang maksimum dari elemen dalam daftar
    private int getMaxLength(List<?> list) {
        int maxLength = 0; // Inisialisasi panjang maksimum
        for (Object item : list) { // Iterasi setiap item dalam daftar
            int length = item.toString().length(); // Hitung panjang string item
            if (length > maxLength) { // Periksa apakah panjang lebih besar dari maxLength saat ini
                maxLength = length; // Perbarui maxLength
            }
        }
        return maxLength; // Kembalikan panjang maksimum
    }

    // Mengambil panjang maksimum untuk kolom harga
    private int getMaxPriceLength() {
        int maxLength = "Harga".length(); // Inisialisasi dengan panjang kata "Harga"
        for (Double price : getProductPrices()) { // Iterasi setiap harga dalam daftar harga produk
            String priceStr = String.format("Rp%.1f", price); // Format harga menjadi string dengan prefiks "Rp"
            if (priceStr.length() > maxLength) { // Periksa apakah panjang string harga lebih besar dari maxLength saat ini
                maxLength = priceStr.length(); // Perbarui maxLength
            }
        }
        return maxLength; // Kembalikan panjang maksimum untuk harga
    }

    // Membuat garis pemisah dengan karakter tertentu
    private String createLine(int length, char ch) {
        StringBuilder sb = new StringBuilder(); // Gunakan StringBuilder untuk efisiensi
        for (int i = 0; i < length; i++) { // Iterasi sebanyak length
            sb.append(ch); // Tambahkan karakter ke StringBuilder
        }
        return sb.toString(); // Kembalikan string hasil
    }

    // Memformat string agar memiliki panjang tetap dengan padding ke kanan
    private String padString(String str, int length) {
        return String.format("%-" + length + "s", str); // Gunakan format string untuk padding
    }

    // Menampilkan tabel produk
    public void displayProducts() {
        if (getProductIds().isEmpty()) { // Periksa apakah daftar produk kosong
            System.out.println("Tidak ada produk yang tersedia."); // Cetak pesan jika kosong
            return; // Keluar dari metode
        }

        // Menghitung lebar maksimum untuk setiap kolom
        int idWidth = Math.max(getMaxLength(getProductIds()), "ID".length());
        int nameWidth = Math.max(getMaxLength(getProductNames()), "Nama".length());
        int categoryWidth = Math.max(getMaxLength(getProductCategories()), "Kategori".length());
        int priceWidth = getMaxPriceLength();
        int photoWidth = Math.max(getMaxLength(getProductPhotos()), "Foto".length());

        // Menghitung total lebar tabel dengan tambahan 13 karakter untuk pemisah
        int totalWidth = idWidth + nameWidth + categoryWidth + priceWidth + photoWidth + 13;

        // Mencetak garis atas tabel
        String headerLine = createLine(totalWidth + 3, '-'); // Buat garis header
        System.out.println(headerLine); // Cetak garis header

        // Mencetak header tabel
        System.out.printf("| %s | %s | %s | %s | %s |\n",
            padString("ID", idWidth),
            padString("Nama", nameWidth),
            padString("Kategori", categoryWidth),
            padString("Harga", priceWidth),
            padString("Foto", photoWidth));
        System.out.println(headerLine); // Cetak garis pemisah header

        // Mencetak isi tabel
        for (int i = 0; i < getProductIds().size(); i++) {
            String priceStr = String.format("Rp%.1f", getProductPrices().get(i)); // Format harga
            System.out.printf("| %s | %s | %s | %s | %s |\n",
                padString(getProductIds().get(i).toString(), idWidth),
                padString(getProductNames().get(i), nameWidth),
                padString(getProductCategories().get(i), categoryWidth),
                padString(priceStr, priceWidth),
                padString(getProductPhotos().get(i), photoWidth));
        }
        System.out.println(headerLine); // Cetak garis bawah tabel
    }


// Metode untuk menambahkan produk baru
public void addProduct(String name, String category, double price, String photo) {
    // Cek jika harga negatif
    if (price < 0) { 
        // Pesan error jika harga negatif
        System.out.println("Harga tidak boleh negatif!"); 
        return; // Menghentikan metode jika harga negatif
    }
    
    // Menambahkan ID produk baru ke dalam daftar ID produk
    getProductIds().add(nextId++);
    
    // Menambahkan nama produk ke dalam daftar nama produk
    getProductNames().add(name);
    
    // Menambahkan kategori produk ke dalam daftar kategori produk
    getProductCategories().add(category);
    
    // Menambahkan harga produk ke dalam daftar harga produk
    getProductPrices().add(price);
    
    // Menambahkan foto produk ke dalam daftar foto produk
    getProductPhotos().add(photo);
    
    // Pesan sukses setelah produk berhasil ditambahkan
    System.out.println("Produk berhasil ditambahkan."); 
}


    // Metode untuk memperbarui produk berdasarkan ID
// Metode untuk memperbarui produk berdasarkan ID
public void updateProduct(int id, String name, String category, double price, String photo) {
    // Mencari indeks produk berdasarkan ID
    int index = getProductIds().indexOf(id);
    
    // Jika produk ditemukan (index != -1)
    if (index != -1) {
        // Memperbarui nama produk pada indeks yang ditemukan
        setProductName(index, name);
        
        // Memperbarui kategori produk pada indeks yang ditemukan
        setProductCategory(index, category);
        
        // Memperbarui harga produk pada indeks yang ditemukan
        setProductPrice(index, price);
        
        // Memperbarui foto produk pada indeks yang ditemukan
        setProductPhoto(index, photo);
        
        // Pesan sukses setelah produk berhasil diperbarui
        System.out.println("Produk berhasil diperbarui.");
    } else {
        // Pesan error jika produk dengan ID tersebut tidak ditemukan
        System.out.println("Produk dengan ID " + id + " tidak ditemukan.");
    }
}

    // Metode untuk menghapus produk berdasarkan ID
// Metode untuk menghapus produk berdasarkan ID
public void deleteProduct(int id) {
    // Mencari indeks produk berdasarkan ID
    int index = getProductIds().indexOf(id);
    
    // Jika produk ditemukan (index != -1)
    if (index != -1) {
        // Menghapus ID produk pada indeks yang ditemukan
        getProductIds().remove(index);
        
        // Menghapus nama produk pada indeks yang ditemukan
        getProductNames().remove(index);
        
        // Menghapus kategori produk pada indeks yang ditemukan
        getProductCategories().remove(index);
        
        // Menghapus harga produk pada indeks yang ditemukan
        getProductPrices().remove(index);
        
        // Menghapus foto produk pada indeks yang ditemukan
        getProductPhotos().remove(index);
        
        // Pesan sukses setelah produk berhasil dihapus
        System.out.println("Produk berhasil dihapus.");
    } else {
        // Pesan error jika produk dengan ID tersebut tidak ditemukan
        System.out.println("Produk dengan ID " + id + " tidak ditemukan.");
    }
}


    // Metode untuk mencari produk berdasarkan nama
public void searchProductByName(String name) {
    // Variabel untuk menandai apakah produk ditemukan
    boolean found = false;
    
    // Daftar untuk menyimpan indeks produk yang cocok
    List<Integer> matchingIndexes = new ArrayList<>();
    
    // Mencari produk dengan nama yang mengandung string pencarian (case-insensitive)
    for (int i = 0; i < getProductNames().size(); i++) {
        // Jika nama produk mengandung nama pencarian, tambahkan indeksnya ke dalam daftar matchingIndexes
        if (getProductNames().get(i).toLowerCase().contains(name.toLowerCase())) {
            matchingIndexes.add(i);
            found = true;
        }
    }

    // Jika tidak ada produk yang ditemukan, tampilkan pesan error
    if (!found) {
        System.out.println("Produk dengan nama '" + name + "' tidak ditemukan.");
        return; // Menghentikan eksekusi jika tidak ada produk yang ditemukan
    }

    // Menghitung panjang kolom berdasarkan panjang data produk yang ada
    int idWidth = Math.max(getMaxLength(getProductIds()), "ID".length());
    int nameWidth = Math.max(getMaxLength(getProductNames()), "Nama".length());
    int categoryWidth = Math.max(getMaxLength(getProductCategories()), "Kategori".length());
    int priceWidth = getMaxPriceLength();
    int photoWidth = Math.max(getMaxLength(getProductPhotos()), "Foto".length());
    
    // Menghitung lebar total untuk mencetak tabel
    int totalWidth = idWidth + nameWidth + categoryWidth + priceWidth + photoWidth + 13;

    // Membuat garis pemisah untuk header tabel
    String headerLine = createLine(totalWidth + 3, '-');
    
    // Menampilkan header hasil pencarian
    System.out.println("\nHasil pencarian untuk '" + name + "':");
    System.out.println(headerLine);
    
    // Menampilkan header kolom
    System.out.printf("| %s | %s | %s | %s | %s |\n",
        padString("ID", idWidth),
        padString("Nama", nameWidth),
        padString("Kategori", categoryWidth),
        padString("Harga", priceWidth),
        padString("Foto", photoWidth));
    
    System.out.println(headerLine);

    // Menampilkan informasi produk yang ditemukan
    for (int index : matchingIndexes) {
        String priceStr = String.format("Rp%.1f", getProductPrices().get(index));
        
        // Menampilkan data produk dalam format tabel
        System.out.printf("| %s | %s | %s | %s | %s |\n",
            padString(getProductIds().get(index).toString(), idWidth),
            padString(getProductNames().get(index), nameWidth),
            padString(getProductCategories().get(index), categoryWidth),
            padString(priceStr, priceWidth),
            padString(getProductPhotos().get(index), photoWidth));
    }
    
    // Menampilkan garis pemisah di bawah tabel
    System.out.println(headerLine);
}


    
    // Metode utama untuk menjalankan program
    public static void main(String[] args) {
        PetShop myPetShop = new PetShop(); // Buat objek PetShop
        Scanner scanner = new Scanner(System.in); // Buat objek Scanner untuk input
        int choice, id; // Variabel untuk pilihan menu dan ID produk
        String name, category, photo; // Variabel untuk nama, kategori, dan foto produk
        double price; // Variabel untuk harga produk

        do {
            // Menampilkan menu
            System.out.println("\n=== SISTEM MANAJEMEN PETSHOP ===");
            System.out.println("1. Tampilkan Data Produk");
            System.out.println("2. Tambah Produk Baru");
            System.out.println("3. Perbarui Produk");
            System.out.println("4. Hapus Produk");
            System.out.println("5. Cari Produk Berdasarkan Nama");
            System.out.println("6. Keluar");
            System.out.print("Pilih menu (1-6): ");
            choice = scanner.nextInt(); // Menerima input pilihan menu

            switch (choice) {
                case 1:
                    myPetShop.displayProducts(); // Tampilkan semua produk
                    break;
                case 2:
                    System.out.print("Masukkan Nama: ");
                    scanner.nextLine(); // Membersihkan buffer
                    name = scanner.nextLine(); // Menerima input nama
                    System.out.print("Masukkan Kategori: ");
                    category = scanner.nextLine(); // Menerima input kategori
                    System.out.print("Masukkan Harga: ");
                    price = scanner.nextDouble(); // Menerima input harga
                    System.out.print("Masukkan Foto: ");
                    scanner.nextLine(); // Membersihkan buffer
                    photo = scanner.nextLine(); // Menerima input foto
                    myPetShop.addProduct(name, category, price, photo); // Tambahkan produk baru
                    break;
                case 3:
                    System.out.print("Masukkan ID Produk: ");
                    id = scanner.nextInt(); // Menerima input ID produk
                    System.out.print("Masukkan Nama: ");
                    scanner.nextLine(); // Membersihkan buffer
                    name = scanner.nextLine(); // Menerima input nama
                    System.out.print("Masukkan Kategori: ");
                    category = scanner.nextLine(); // Menerima input kategori
                    System.out.print("Masukkan Harga: ");
                    price = scanner.nextDouble(); // Menerima input harga
                    System.out.print("Masukkan Foto: ");
                    scanner.nextLine(); // Membersihkan buffer
                    photo = scanner.nextLine(); // Menerima input foto
                    myPetShop.updateProduct(id, name, category, price, photo); // Perbarui produk
                    break;
                case 4:
                    System.out.print("Masukkan ID Produk: ");
                    id = scanner.nextInt(); // Menerima input ID produk
                    myPetShop.deleteProduct(id); // Hapus produk
                    break;
                case 5:
                    System.out.print("Masukkan Nama Produk: ");
                    scanner.nextLine(); // Membersihkan buffer
                    name = scanner.nextLine(); // Menerima input nama
                    myPetShop.searchProductByName(name); // Cari produk berdasarkan nama
                    break;
                case 6:
                    System.out.println("Terima kasih!"); // Pesan keluar
                    break;
                default:
                    System.out.println("Pilihan tidak valid!"); // Pesan error untuk pilihan tidak valid
            }
        } while (choice != 6); // Ulangi sampai pengguna memilih keluar

        scanner.close(); // Tutup Scanner
    }
}